In [1]:
import pandas as pd
import numpy as np

In [2]:
#import data
df1 = pd.read_csv('marketing_customer_analysis.csv')

In [3]:
df1

Unnamed: 0 Customer       State  Customer Lifetime Value Response  \
0               0  DK49336     Arizona              4809.216960       No   
1               1  KX64629  California              2228.525238       No   
2               2  LZ68649  Washington             14947.917300       No   
3               3  XL78013      Oregon             22332.439460      Yes   
4               4  QA50777      Oregon              9025.067525       No   
...           ...      ...         ...                      ...      ...   
10905       10905  FE99816      Nevada             15563.369440       No   
10906       10906  KX53892      Oregon              5259.444853       No   
10907       10907  TL39050     Arizona             23893.304100       No   
10908       10908  WA60547  California             11971.977650       No   
10909       10909  IV32877         NaN              6857.519928      NaN   

       Coverage Education Effective To Date EmploymentStatus Gender  ...  \
0         Basic   College           2/18/11         Employed      M  ...   
1         Basic   College           1/18/11       Unemployed      F  ...   
2         Basic  Bachelor           2/10/11         Employed      M  ...   
3      Extended   College           1/11/11         Employed      M  ...   
4       Premium  Bachelor           1/17/11    Medical Leave      F  ...   
...         ...       ...               ...              ...    ...  ...   
10905   Premium  Bachelor           1/19/11       Unemployed      F  ...   
10906     Basic   College            1/6/11         Employed      F  ...   
10907  Extended  Bachelor            2/6/11         Employed      F  ...   
10908   Premium   College           2/13/11         Employed      F  ...   
10909     Basic  Bachelor            1/8/11       Unemployed      M  ...   

       Number of Open Complaints Number of Policies     Policy Type  \
0                            0.0                  9  Corporate Auto   
1                            0.0                  1   Personal Auto   
2                            0.0                  2   Personal Auto   
3                            0.0                  2  Corporate Auto   
4                            NaN                  7   Personal Auto   
...                          ...                ...             ...   
10905                        NaN                  7   Personal Auto   
10906                        0.0                  6   Personal Auto   
10907                        0.0                  2  Corporate Auto   
10908                        4.0                  6   Personal Auto   
10909                        0.0                  3   Personal Auto   

             Policy  Renew Offer Type  Sales Channel  Total Claim Amount  \
0      Corporate L3            Offer3          Agent          292.800000   
1       Personal L3            Offer4    Call Center          744.924331   
2       Personal L3            Offer3    Call Center          480.000000   
3      Corporate L3            Offer2         Branch          484.013411   
4       Personal L2            Offer1         Branch          707.925645   
...             ...               ...            ...                 ...   
10905   Personal L1            Offer3            Web         1214.400000   
10906   Personal L3            Offer2         Branch          273.018929   
10907  Corporate L3            Offer1            Web          381.306996   
10908   Personal L1            Offer1         Branch          618.288849   
10909   Personal L1            Offer4            Web         1021.719397   

       Vehicle Class Vehicle Size Vehicle Type  
0      Four-Door Car      Medsize          NaN  
1      Four-Door Car      Medsize          NaN  
2                SUV      Medsize            A  
3      Four-Door Car      Medsize            A  
4      Four-Door Car      Medsize          NaN  
...              ...          ...          ...  
10905     Luxury Car      Medsize            A  
10906  Four-Door Car   

In [4]:
#Here is the shape:
df1.shape

(10910, 26)

In [5]:
df1.columns

Index(['Unnamed: 0', 'Customer', 'State', 'Customer Lifetime Value',
       'Response', 'Coverage', 'Education', 'Effective To Date',
       'EmploymentStatus', 'Gender', 'Income', 'Location Code',
       'Marital Status', 'Monthly Premium Auto', 'Months Since Last Claim',
       'Months Since Policy Inception', 'Number of Open Complaints',
       'Number of Policies', 'Policy Type', 'Policy', 'Renew Offer Type',
       'Sales Channel', 'Total Claim Amount', 'Vehicle Class', 'Vehicle Size',
       'Vehicle Type'],
      dtype='object')

In [6]:
df1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10910 entries, 0 to 10909
Data columns (total 26 columns):
 #   Column                         Non-Null Count  Dtype  
---  ------                         --------------  -----  
 0   Unnamed: 0                     10910 non-null  int64  
 1   Customer                       10910 non-null  object 
 2   State                          10279 non-null  object 
 3   Customer Lifetime Value        10910 non-null  float64
 4   Response                       10279 non-null  object 
 5   Coverage                       10910 non-null  object 
 6   Education                      10910 non-null  object 
 7   Effective To Date              10910 non-null  object 
 8   EmploymentStatus               10910 non-null  object 
 9   Gender                         10910 non-null  object 
 10  Income                         10910 non-null  int64  
 11  Location Code                  10910 non-null  object 
 12  Marital Status                 10910 non-null 

In [7]:
#New names for columns
cols = [col_name.upper().replace(' ', '_') for col_name in df1.columns]
print(cols)

['UNNAMED:_0', 'CUSTOMER', 'STATE', 'CUSTOMER_LIFETIME_VALUE', 'RESPONSE', 'COVERAGE', 'EDUCATION', 'EFFECTIVE_TO_DATE', 'EMPLOYMENTSTATUS', 'GENDER', 'INCOME', 'LOCATION_CODE', 'MARITAL_STATUS', 'MONTHLY_PREMIUM_AUTO', 'MONTHS_SINCE_LAST_CLAIM', 'MONTHS_SINCE_POLICY_INCEPTION', 'NUMBER_OF_OPEN_COMPLAINTS', 'NUMBER_OF_POLICIES', 'POLICY_TYPE', 'POLICY', 'RENEW_OFFER_TYPE', 'SALES_CHANNEL', 'TOTAL_CLAIM_AMOUNT', 'VEHICLE_CLASS', 'VEHICLE_SIZE', 'VEHICLE_TYPE']


In [8]:
#Renamed the column names
df1.columns = cols

In [9]:
df1.head(5)

UNNAMED:_0 CUSTOMER       STATE  CUSTOMER_LIFETIME_VALUE RESPONSE  \
0           0  DK49336     Arizona              4809.216960       No   
1           1  KX64629  California              2228.525238       No   
2           2  LZ68649  Washington             14947.917300       No   
3           3  XL78013      Oregon             22332.439460      Yes   
4           4  QA50777      Oregon              9025.067525       No   

   COVERAGE EDUCATION EFFECTIVE_TO_DATE EMPLOYMENTSTATUS GENDER  ...  \
0     Basic   College           2/18/11         Employed      M  ...   
1     Basic   College           1/18/11       Unemployed      F  ...   
2     Basic  Bachelor           2/10/11         Employed      M  ...   
3  Extended   College           1/11/11         Employed      M  ...   
4   Premium  Bachelor           1/17/11    Medical Leave      F  ...   

   NUMBER_OF_OPEN_COMPLAINTS NUMBER_OF_POLICIES     POLICY_TYPE        POLICY  \
0                        0.0                  9  Corporate Auto  Corporate L3   
1                        0.0                  1   Personal Auto   Personal L3   
2                        0.0                  2   Personal Auto   Personal L3   
3                        0.0                  2  Corporate Auto  Corporate L3   
4                        NaN                  7   Personal Auto   Personal L2   

   RENEW_OFFER_TYPE  SALES_CHANNEL  TOTAL_CLAIM_AMOUNT  VEHICLE_CLASS  \
0            Offer3          Agent          292.800000  Four-Door Car   
1            Offer4    Call Center          744.924331  Four-Door Car   
2            Offer3    Call Center          480.000000            SUV   
3            Offer2         Branch          484.013411  Four-Door Car   
4            Offer1         Branch          707.925645  Four-Door Car   

  VEHICLE_SIZE VEHICLE_TYPE  
0      Medsize          NaN  
1      Medsize          NaN  
2      Medsize            A  
3      Medsize            A  
4      Medsize          NaN  

[5 rows x 26 columns]

# #Here are if the variables are numerical or categorical
 0   Unnamed: 0                     10910 non-null  int64  NUMERICAL
 1   Customer                       10910 non-null  object CATEGORICAL
 2   State                          10279 non-null  object CATEGORICAL
 3   Customer Lifetime Value        10910 non-null  float64NUMERICAL
 4   Response                       10279 non-null  object CATEGORICAL
 5   Coverage                       10910 non-null  object CATEGORICAL
 6   Education                      10910 non-null  object CATEGORICAL
 7   Effective To Date              10910 non-null  object DATE
 8   EmploymentStatus               10910 non-null  object CATEGORICAL
 9   Gender                         10910 non-null  object CATEGORICAL
 10  Income                         10910 non-null  int64  CATEGORICAL
 11  Location Code                  10910 non-null  object CATEGORICAL
 12  Marital Status                 10910 non-null  object CATEGORICAL
 13  Monthly Premium Auto           10910 non-null  int64  NUMERICAL
 14  Months Since Last Claim        10277 non-null  float64NUMERICAL
 15  Months Since Policy Inception  10910 non-null  int64  NUMERICAL
 16  Number of Open Complaints      10277 non-null  float64NUMERICAL
 17  Number of Policies             10910 non-null  int64  NUMERICAL
 18  Policy Type                    10910 non-null  object CATEGORICAL
 19  Policy                         10910 non-null  object CATEGORICAL
 20  Renew Offer Type               10910 non-null  object CATEGORICAL
 21  Sales Channel                  10910 non-null  object CATEGORICAL
 22  Total Claim Amount             10910 non-null  float64NUMERICAL
 23  Vehicle Class                  10288 non-null  object CATEGORICAL
 24  Vehicle Size                   10288 non-null  object CATEGORICAL
 25  Vehicle Type                   5428 non-null   object CATEGORICAL

In [10]:
#How many NAN are in each columns
df1.isna().sum()

UNNAMED:_0                          0
CUSTOMER                            0
STATE                             631
CUSTOMER_LIFETIME_VALUE             0
RESPONSE                          631
COVERAGE                            0
EDUCATION                           0
EFFECTIVE_TO_DATE                   0
EMPLOYMENTSTATUS                    0
GENDER                              0
INCOME                              0
LOCATION_CODE                       0
MARITAL_STATUS                      0
MONTHLY_PREMIUM_AUTO                0
MONTHS_SINCE_LAST_CLAIM           633
MONTHS_SINCE_POLICY_INCEPTION       0
NUMBER_OF_OPEN_COMPLAINTS         633
NUMBER_OF_POLICIES                  0
POLICY_TYPE                         0
POLICY                              0
RENEW_OFFER_TYPE                    0
SALES_CHANNEL                       0
TOTAL_CLAIM_AMOUNT                  0
VEHICLE_CLASS                     622
VEHICLE_SIZE                      622
VEHICLE_TYPE                     5482
dtype: int64

In [11]:
#Thanks to ISI for the idea to separate the numeric and Strings variables
#Select numeric columns
df_numeric=df1.select_dtypes(include=["float", "int"]) 
null_columns = df_numeric.columns[df_numeric.isna().sum() != 0].tolist()
print(null_columns)
#Put 0 in those which have nulls.
for  Column_to_change in null_columns:
    df_numeric[Column_to_change] = df_numeric[Column_to_change].fillna(0) 
df_numeric.info()

['MONTHS_SINCE_LAST_CLAIM', 'NUMBER_OF_OPEN_COMPLAINTS']
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10910 entries, 0 to 10909
Data columns (total 9 columns):
 #   Column                         Non-Null Count  Dtype  
---  ------                         --------------  -----  
 0   UNNAMED:_0                     10910 non-null  int64  
 1   CUSTOMER_LIFETIME_VALUE        10910 non-null  float64
 2   INCOME                         10910 non-null  int64  
 3   MONTHLY_PREMIUM_AUTO           10910 non-null  int64  
 4   MONTHS_SINCE_LAST_CLAIM        10910 non-null  float64
 5   MONTHS_SINCE_POLICY_INCEPTION  10910 non-null  int64  
 6   NUMBER_OF_OPEN_COMPLAINTS      10910 non-null  float64
 7   NUMBER_OF_POLICIES             10910 non-null  int64  
 8   TOTAL_CLAIM_AMOUNT             10910 non-null  float64
dtypes: float64(4), int64(5)
memory usage: 767.2 KB


In [12]:
#Select Object Variables
df_object=df1.select_dtypes(include=["object"]) 
null_columns = df_object.columns[df_object.isna().sum() != 0].tolist()
print(null_columns)
#Put Unkwon in those which have nulls.
for  Column_to_change in null_columns:
    df_object[Column_to_change] = df_object[Column_to_change].fillna('Unknown') 
df_object.info()

['STATE', 'RESPONSE', 'VEHICLE_CLASS', 'VEHICLE_SIZE', 'VEHICLE_TYPE']
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10910 entries, 0 to 10909
Data columns (total 17 columns):
 #   Column             Non-Null Count  Dtype 
---  ------             --------------  ----- 
 0   CUSTOMER           10910 non-null  object
 1   STATE              10910 non-null  object
 2   RESPONSE           10910 non-null  object
 3   COVERAGE           10910 non-null  object
 4   EDUCATION          10910 non-null  object
 5   EFFECTIVE_TO_DATE  10910 non-null  object
 6   EMPLOYMENTSTATUS   10910 non-null  object
 7   GENDER             10910 non-null  object
 8   LOCATION_CODE      10910 non-null  object
 9   MARITAL_STATUS     10910 non-null  object
 10  POLICY_TYPE        10910 non-null  object
 11  POLICY             10910 non-null  object
 12  RENEW_OFFER_TYPE   10910 non-null  object
 13  SALES_CHANNEL      10910 non-null  object
 14  VEHICLE_CLASS      10910 non-null  object
 15  VEHICLE_SIZE    

In [14]:
#Here is the final dataframe already cleaned
df_clean=pd.concat([df_numeric,df_object],axis=1)
df_clean.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10910 entries, 0 to 10909
Data columns (total 26 columns):
 #   Column                         Non-Null Count  Dtype  
---  ------                         --------------  -----  
 0   UNNAMED:_0                     10910 non-null  int64  
 1   CUSTOMER_LIFETIME_VALUE        10910 non-null  float64
 2   INCOME                         10910 non-null  int64  
 3   MONTHLY_PREMIUM_AUTO           10910 non-null  int64  
 4   MONTHS_SINCE_LAST_CLAIM        10910 non-null  float64
 5   MONTHS_SINCE_POLICY_INCEPTION  10910 non-null  int64  
 6   NUMBER_OF_OPEN_COMPLAINTS      10910 non-null  float64
 7   NUMBER_OF_POLICIES             10910 non-null  int64  
 8   TOTAL_CLAIM_AMOUNT             10910 non-null  float64
 9   CUSTOMER                       10910 non-null  object 
 10  STATE                          10910 non-null  object 
 11  RESPONSE                       10910 non-null  object 
 12  COVERAGE                       10910 non-null 

In [15]:
df_clean.isna().sum()

UNNAMED:_0                       0
CUSTOMER_LIFETIME_VALUE          0
INCOME                           0
MONTHLY_PREMIUM_AUTO             0
MONTHS_SINCE_LAST_CLAIM          0
MONTHS_SINCE_POLICY_INCEPTION    0
NUMBER_OF_OPEN_COMPLAINTS        0
NUMBER_OF_POLICIES               0
TOTAL_CLAIM_AMOUNT               0
CUSTOMER                         0
STATE                            0
RESPONSE                         0
COVERAGE                         0
EDUCATION                        0
EFFECTIVE_TO_DATE                0
EMPLOYMENTSTATUS                 0
GENDER                           0
LOCATION_CODE                    0
MARITAL_STATUS                   0
POLICY_TYPE                      0
POLICY                           0
RENEW_OFFER_TYPE                 0
SALES_CHANNEL                    0
VEHICLE_CLASS                    0
VEHICLE_SIZE                     0
VEHICLE_TYPE                     0
dtype: int64

In [16]:
#Change the data type for the date 
df_time=df_clean
df_time['EFFECTIVE_TO_DATE'] = pd.to_datetime(df_time['EFFECTIVE_TO_DATE'])

In [17]:
df_time.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10910 entries, 0 to 10909
Data columns (total 26 columns):
 #   Column                         Non-Null Count  Dtype         
---  ------                         --------------  -----         
 0   UNNAMED:_0                     10910 non-null  int64         
 1   CUSTOMER_LIFETIME_VALUE        10910 non-null  float64       
 2   INCOME                         10910 non-null  int64         
 3   MONTHLY_PREMIUM_AUTO           10910 non-null  int64         
 4   MONTHS_SINCE_LAST_CLAIM        10910 non-null  float64       
 5   MONTHS_SINCE_POLICY_INCEPTION  10910 non-null  int64         
 6   NUMBER_OF_OPEN_COMPLAINTS      10910 non-null  float64       
 7   NUMBER_OF_POLICIES             10910 non-null  int64         
 8   TOTAL_CLAIM_AMOUNT             10910 non-null  float64       
 9   CUSTOMER                       10910 non-null  object        
 10  STATE                          10910 non-null  object        
 11  RESPONSE       

In [18]:
#Extract month from date
df_time['MONTH'] = df_time['EFFECTIVE_TO_DATE'].dt.month
df_time

UNNAMED:_0  CUSTOMER_LIFETIME_VALUE  INCOME  MONTHLY_PREMIUM_AUTO  \
0               0              4809.216960   48029                    61   
1               1              2228.525238       0                    64   
2               2             14947.917300   22139                   100   
3               3             22332.439460   49078                    97   
4               4              9025.067525   23675                   117   
...           ...                      ...     ...                   ...   
10905       10905             15563.369440       0                   253   
10906       10906              5259.444853   61146                    65   
10907       10907             23893.304100   39837                   201   
10908       10908             11971.977650   64195                   158   
10909       10909              6857.519928       0                   101   

       MONTHS_SINCE_LAST_CLAIM  MONTHS_SINCE_POLICY_INCEPTION  \
0                          7.0                             52   
1                          3.0                             26   
2                         34.0                             31   
3                         10.0                              3   
4                          0.0                             31   
...                        ...                            ...   
10905                      0.0                             40   
10906                      7.0                             68   
10907                     11.0                             63   
10908                      0.0                             27   
10909                     31.0                              1   

       NUMBER_OF_OPEN_COMPLAINTS  NUMBER_OF_POLICIES  TOTAL_CLAIM_AMOUNT  \
0                            0.0                   9          292.800000   
1                            0.0                   1          744.924331   
2                            0.0                   2          480.000000   
3                            0.0                   2          484.013411   
4                            0.0                   7          707.925645   
...                          ...                 ...                 ...   
10905                        0.0                   7         1214.400000   
10906                        0.0                   6          273.018929   
10907                        0.0                   2          381.306996   
10908                        4.0                   6          618.288849   
10909                        0.0                   3         1021.719397   

      CUSTOMER  ... LOCATION_CODE MARITAL_STATUS     POLICY_TYPE  \
0      DK49336  ...      Suburban        Married  Corporate Auto   
1      KX64629  ...      Suburban         Single   Personal Auto   
2      LZ68649  ...      Suburban         Single   Personal Auto   
3      XL78013  ...      Suburban         Single  Corporate Auto   
4      QA50777  ...      Suburban        Married   Personal Auto   
...        ...  ...           ...            ...             ...   
10905  FE99816  ...      Suburban        Married   Personal Auto   
10906  KX53892  ...         Urban        Married   Personal Auto   
10907  TL39050  ...         Rural        Married  Corporate Auto   
10908  WA60547  ...         Urban       Divorced   Personal Auto   
10909  IV32877  ...      Suburban         Single   Personal Auto   

             POLICY RENEW_OFFER_TYPE SALES_CHANNEL  VEHICLE_CLASS  \
0      Corporate L3           Offer3         Agent  Four-Door Car   
1       Personal L3           Offer4   Call Center  Four-Door Car   
2       Personal L3           Offer3   Call Center            SUV   
3      Corporate L3           Offer2        Branch  Four-Door Car   
4       Personal L2           Offer1        Branch  Four-Door Car   
...             ...              ...           ...            ...   
10905   Personal L1           Offer3           Web     Luxury Car   
10906   Personal L3           O

In [ ]:
#explore dates
df_time_2 = df_time[df_time['EFFECTIVE_TO_DATE'] < '2011-01-01' ]
df_time_2

In [19]:
df_time['MONTH'].value_counts()

1    5818
2    5092
Name: MONTH, dtype: int64

In [20]:
#Filter only records in First quarter. Also only months 1,2 and 3
df_time_2=df_time[df_time['MONTH'].isin([1,2,3])]
df_time_2

UNNAMED:_0  CUSTOMER_LIFETIME_VALUE  INCOME  MONTHLY_PREMIUM_AUTO  \
0               0              4809.216960   48029                    61   
1               1              2228.525238       0                    64   
2               2             14947.917300   22139                   100   
3               3             22332.439460   49078                    97   
4               4              9025.067525   23675                   117   
...           ...                      ...     ...                   ...   
10905       10905             15563.369440       0                   253   
10906       10906              5259.444853   61146                    65   
10907       10907             23893.304100   39837                   201   
10908       10908             11971.977650   64195                   158   
10909       10909              6857.519928       0                   101   

       MONTHS_SINCE_LAST_CLAIM  MONTHS_SINCE_POLICY_INCEPTION  \
0                          7.0                             52   
1                          3.0                             26   
2                         34.0                             31   
3                         10.0                              3   
4                          0.0                             31   
...                        ...                            ...   
10905                      0.0                             40   
10906                      7.0                             68   
10907                     11.0                             63   
10908                      0.0                             27   
10909                     31.0                              1   

       NUMBER_OF_OPEN_COMPLAINTS  NUMBER_OF_POLICIES  TOTAL_CLAIM_AMOUNT  \
0                            0.0                   9          292.800000   
1                            0.0                   1          744.924331   
2                            0.0                   2          480.000000   
3                            0.0                   2          484.013411   
4                            0.0                   7          707.925645   
...                          ...                 ...                 ...   
10905                        0.0                   7         1214.400000   
10906                        0.0                   6          273.018929   
10907                        0.0                   2          381.306996   
10908                        4.0                   6          618.288849   
10909                        0.0                   3         1021.719397   

      CUSTOMER  ... LOCATION_CODE MARITAL_STATUS     POLICY_TYPE  \
0      DK49336  ...      Suburban        Married  Corporate Auto   
1      KX64629  ...      Suburban         Single   Personal Auto   
2      LZ68649  ...      Suburban         Single   Personal Auto   
3      XL78013  ...      Suburban         Single  Corporate Auto   
4      QA50777  ...      Suburban        Married   Personal Auto   
...        ...  ...           ...            ...             ...   
10905  FE99816  ...      Suburban        Married   Personal Auto   
10906  KX53892  ...         Urban        Married   Personal Auto   
10907  TL39050  ...         Rural        Married  Corporate Auto   
10908  WA60547  ...         Urban       Divorced   Personal Auto   
10909  IV32877  ...      Suburban         Single   Personal Auto   

             POLICY RENEW_OFFER_TYPE SALES_CHANNEL  VEHICLE_CLASS  \
0      Corporate L3           Offer3         Agent  Four-Door Car   
1       Personal L3           Offer4   Call Center  Four-Door Car   
2       Personal L3           Offer3   Call Center            SUV   
3      Corporate L3           Offer2        Branch  Four-Door Car   
4       Personal L2           Offer1        Branch  Four-Door Car   
...             ...              ...           ...            ...   
10905   Personal L1           Offer3           Web     Luxury Car   
10906   Personal L3           O

# Here starts the extra excercise

In [23]:
#Function to clean the data
def DATA_FRAME_CLEAN(data_frame_input):
    #Standarize column names
    cols = [col_name.upper().replace(' ', '_') for col_name in data_frame_input.columns]
    data_frame_input.columns = cols
    
    #Thanks to ISI for the idea to separate the numeric and Strings variables
    #Clean nulls
    #Select numeric columns
    df_numeric=data_frame_input.select_dtypes(include=["float", "int"]) 
    null_columns = df_numeric.columns[df_numeric.isna().sum() != 0].tolist()
    print(null_columns)
    #Put 0 in those which have nulls.
    for  Column_to_change in null_columns:
        df_numeric[Column_to_change] = df_numeric[Column_to_change].fillna(0) 
    
    #Select Object Variables
    df_object=data_frame_input.select_dtypes(include=[object]) 
    null_columns = df_object.columns[df_object.isna().sum() != 0].tolist()
    print(null_columns)
    #Put Unkwon in those which have nulls.
    for  Column_to_change in null_columns:
        df_object[Column_to_change] = df_object[Column_to_change].fillna('Unknown') 
    
    #Merge both branches
    data_frame_input=pd.concat([df_numeric,df_object],axis=1)
    print(data_frame_input.columns)
    
    #Change EFFECTIVE_TO_DATE to date type
    data_frame_input['EFFECTIVE_TO_DATE'] = pd.to_datetime(data_frame_input['EFFECTIVE_TO_DATE'])
    
    #Assign to ouput
    data_frame_output=data_frame_input
    
    return data_frame_output

In [24]:
#Input the dataframe to clean
data_frame_input = pd.read_csv('marketing_customer_analysis.csv')
data_good=DATA_FRAME_CLEAN(data_frame_input)
data_good

['MONTHS_SINCE_LAST_CLAIM', 'NUMBER_OF_OPEN_COMPLAINTS']
['STATE', 'RESPONSE', 'VEHICLE_CLASS', 'VEHICLE_SIZE', 'VEHICLE_TYPE']
Index(['UNNAMED:_0', 'CUSTOMER_LIFETIME_VALUE', 'INCOME',
       'MONTHLY_PREMIUM_AUTO', 'MONTHS_SINCE_LAST_CLAIM',
       'MONTHS_SINCE_POLICY_INCEPTION', 'NUMBER_OF_OPEN_COMPLAINTS',
       'NUMBER_OF_POLICIES', 'TOTAL_CLAIM_AMOUNT', 'CUSTOMER', 'STATE',
       'RESPONSE', 'COVERAGE', 'EDUCATION', 'EFFECTIVE_TO_DATE',
       'EMPLOYMENTSTATUS', 'GENDER', 'LOCATION_CODE', 'MARITAL_STATUS',
       'POLICY_TYPE', 'POLICY', 'RENEW_OFFER_TYPE', 'SALES_CHANNEL',
       'VEHICLE_CLASS', 'VEHICLE_SIZE', 'VEHICLE_TYPE'],
      dtype='object')


UNNAMED:_0  CUSTOMER_LIFETIME_VALUE  INCOME  MONTHLY_PREMIUM_AUTO  \
0               0              4809.216960   48029                    61   
1               1              2228.525238       0                    64   
2               2             14947.917300   22139                   100   
3               3             22332.439460   49078                    97   
4               4              9025.067525   23675                   117   
...           ...                      ...     ...                   ...   
10905       10905             15563.369440       0                   253   
10906       10906              5259.444853   61146                    65   
10907       10907             23893.304100   39837                   201   
10908       10908             11971.977650   64195                   158   
10909       10909              6857.519928       0                   101   

       MONTHS_SINCE_LAST_CLAIM  MONTHS_SINCE_POLICY_INCEPTION  \
0                          7.0                             52   
1                          3.0                             26   
2                         34.0                             31   
3                         10.0                              3   
4                          0.0                             31   
...                        ...                            ...   
10905                      0.0                             40   
10906                      7.0                             68   
10907                     11.0                             63   
10908                      0.0                             27   
10909                     31.0                              1   

       NUMBER_OF_OPEN_COMPLAINTS  NUMBER_OF_POLICIES  TOTAL_CLAIM_AMOUNT  \
0                            0.0                   9          292.800000   
1                            0.0                   1          744.924331   
2                            0.0                   2          480.000000   
3                            0.0                   2          484.013411   
4                            0.0                   7          707.925645   
...                          ...                 ...                 ...   
10905                        0.0                   7         1214.400000   
10906                        0.0                   6          273.018929   
10907                        0.0                   2          381.306996   
10908                        4.0                   6          618.288849   
10909                        0.0                   3         1021.719397   

      CUSTOMER  ... GENDER LOCATION_CODE MARITAL_STATUS     POLICY_TYPE  \
0      DK49336  ...      M      Suburban        Married  Corporate Auto   
1      KX64629  ...      F      Suburban         Single   Personal Auto   
2      LZ68649  ...      M      Suburban         Single   Personal Auto   
3      XL78013  ...      M      Suburban         Single  Corporate Auto   
4      QA50777  ...      F      Suburban        Married   Personal Auto   
...        ...  ...    ...           ...            ...             ...   
10905  FE99816  ...      F      Suburban        Married   Personal Auto   
10906  KX53892  ...      F         Urban        Married   Personal Auto   
10907  TL39050  ...      F         Rural        Married  Corporate Auto   
10908  WA60547  ...      F         Urban       Divorced   Personal Auto   
10909  IV32877  ...      M      Suburban         Single   Personal Auto   

             POLICY RENEW_OFFER_TYPE SALES_CHANNEL  VEHICLE_CLASS  \
0      Corporate L3           Offer3         Agent  Four-Door Car   
1       Personal L3           Offer4   Call Center  Four-Door Car   
2       Personal L3           Offer3   Call Center            SUV   
3      Corporate L3           Offer2        Branch  Four-Door Car   
4       Personal L2           Offer1        Branch  Four-Door Car   
...             ...              ...           ...            ...   
10905   Personal